staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1175 @@
+				"int(val3) == val1",
+				"int(val4) == val1",
+				"int(val6) == val1",
+				"double(val1) == val2",
+				"double(val6) == val2",
+				"bytes(val6) == val7",
+				"string(val1) == val6",
+				"string(val2) == '10'",
+				"string(val3) == '10.49'",
+				"string(val4) == '10.5'",
+				"string(val5) == 'true'",
+				"string(val7) == val6",
+			},
+		},
+		{name: "lists",
+			obj:    objs([]interface{}{1, 2, 3}, []interface{}{1, 2, 3}),
+			schema: schemas(listType(&integerType), listType(&integerType)),
+			valid: []string{
+				ValsEqualThemselvesAndDataLiteral("val1", "val2", "[1, 2, 3]"),
+				"1 in val1",
+				"val2[0] in val1",
+				"!(0 in val1)",
+			},
+		},
+		{name: "listSets",
+			obj:    objs([]interface{}{"a", "b", "c"}, []interface{}{"a", "c", "b"}),
+			schema: schemas(listSetType(&stringType), listSetType(&stringType)),
+			valid: []string{
+				// equal even though order is different
+				"val1 == ['c', 'b', 'a']",
+				"val1 == val2",
+				"'a' in val1",
+				"val2[0] in val1",
+				"!('x' in val1)",
+			},
+		},
+		{name: "listMaps",
+			obj: map[string]interface{}{
+				"objs": []interface{}{
+					[]interface{}{
+						map[string]interface{}{"k": "a", "v": '1'},
+						map[string]interface{}{"k": "b", "v": '2'},
+					},
+					[]interface{}{
+						map[string]interface{}{"k": "b", "v": '2'},
+						map[string]interface{}{"k": "a", "v": '1'},
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"objs": listType(listMapTypePtr([]string{"k"}, objectTypePtr(map[string]schema.Structural{
+					"k": stringType,
+					"v": stringType,
+				}))),
+			}),
+			valid: []string{
+				"objs[0] == objs[1]", // equal even though order is different
+			},
+			errors: map[string]string{
+				"objs[0] == {'k': 'a', 'v': '1'}": "no matching overload for '_==_'", // objects cannot be compared against a data literal map
+			},
+		},
+		{name: "maps",
+			obj:    objs(map[string]interface{}{"k1": "a", "k2": 'b'}, map[string]interface{}{"k2": 'b', "k1": "a"}),
+			schema: schemas(mapType(&stringType), mapType(&stringType)),
+			valid: []string{
+				"val1 == val2", // equal even though order is different
+				"'k1' in val1",
+				"!('k3' in val1)",
+			},
+		},
+		{name: "objects",
+			obj: map[string]interface{}{
+				"objs": []interface{}{
+					map[string]interface{}{"f1": "a", "f2": "b"},
+					map[string]interface{}{"f1": "a", "f2": "b"},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"objs": listType(objectTypePtr(map[string]schema.Structural{
+					"f1": stringType,
+					"f2": stringType,
+				})),
+			}),
+			valid: []string{
+				"objs[0] == objs[1]",
+			},
+			errors: map[string]string{
+				"objs[0] == {'f1': 'a', 'f2': 'b'}": "no matching overload for '_==_'", // objects cannot be compared against a data literal map
+			},
+		},
+		{name: "object access",
+			obj: map[string]interface{}{
+				"a": map[string]interface{}{
+					"b": 1,
+					"d": 2,
```

Comment:
wouldn't it be better to set "d" to null to verify that it is preserved with nullable set to true?

Reply:
I've made d null here. There is a larger set of nullable tests below that also explain the behavior. Specifically, we've decided to treat a null valued field the same as an unset field in CEL.